File: Xero.WebApi/Controllers/RunDefinitionsController.cs
using Microsoft.AspNetCore.Mvc;
using Xero.WebApi.Models;
using Xero.WebApi.Services;

namespace Xero.WebApi.Controllers;

/// <summary>
/// CRUD for reusable NRT run definitions, plus an execute endpoint that creates a run execution from a definition.
/// </summary>
[ApiController]
[Route("api/run-definitions")]
[Produces("application/json")]
public sealed class RunDefinitionsController : ControllerBase
{
    private readonly INrtRunDefinitionService _definitionService;
    private readonly INrtService              _nrtService;

    public RunDefinitionsController(
        INrtRunDefinitionService definitionService,
        INrtService              nrtService)
    {
        _definitionService = definitionService;
        _nrtService        = nrtService;
    }

    /// <summary>
    /// Returns all active run definitions (summary view, no nested DB config).
    /// </summary>
    [HttpGet]
    [ProducesResponseType(typeof(IReadOnlyList<NrtRunDefinitionSummary>), StatusCodes.Status200OK)]
    public async Task<ActionResult<IReadOnlyList<NrtRunDefinitionSummary>>> GetAll(CancellationToken ct)
    {
        var items = await _definitionService.GetAllAsync(ct);
        return Ok(items);
    }

    /// <summary>
    /// Returns the full detail for a single run definition, including DB connection settings and column schema.
    /// </summary>
    [HttpGet("{id:guid}")]
    [ProducesResponseType(typeof(NrtRunDefinitionDto), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<NrtRunDefinitionDto>> GetById(Guid id, CancellationToken ct)
    {
        var def = await _definitionService.GetByIdAsync(id, ct);
        if (def is null)
            return NotFound($"Definition {id} not found.");
        return Ok(def);
    }

    /// <summary>
    /// Creates a new run definition.
    /// </summary>
    /// <response code="201">Definition created. Location header points to the new resource.</response>
    [HttpPost]
    [ProducesResponseType(typeof(object), StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> Create(
        [FromBody] SaveNrtRunDefinitionRequest req,
        CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(req.Name))
            return BadRequest("Name is required.");

        var id = await _definitionService.CreateAsync(req, ct);
        return CreatedAtAction(nameof(GetById), new { id }, new { definitionId = id });
    }

    /// <summary>
    /// Updates an existing run definition.
    /// </summary>
    /// <response code="204">Definition updated successfully.</response>
    /// <response code="404">Definition not found or has been deleted.</response>
    [HttpPut("{id:guid}")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> Update(
        Guid id,
        [FromBody] SaveNrtRunDefinitionRequest req,
        CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(req.Name))
            return BadRequest("Name is required.");

        var updated = await _definitionService.UpdateAsync(id, req, ct);
        if (!updated)
            return NotFound($"Definition {id} not found.");
        return NoContent();
    }

    /// <summary>
    /// Soft-deletes a run definition (sets <c>is_active = false</c>).
    /// Linked run executions are preserved.
    /// </summary>
    /// <response code="204">Definition deleted.</response>
    /// <response code="404">Definition not found or already deleted.</response>
    [HttpDelete("{id:guid}")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> Delete(Guid id, CancellationToken ct)
    {
        var deleted = await _definitionService.DeleteAsync(id, ct);
        if (!deleted)
            return NotFound($"Definition {id} not found.");
        return NoContent();
    }

    /// <summary>
    /// Executes the NRT pipeline from a saved definition.
    /// Only <c>ValuationDate</c> is required from the caller — all other parameters come from the definition.
    /// Returns a full <see cref="RunExecutionResponse"/> with status, counts, and pass/fail result.
    /// </summary>
    /// <response code="200">Run completed — inspect <c>Passed</c> and diff counts.</response>
    /// <response code="404">Definition not found.</response>
    [HttpPost("{id:guid}/execute")]
    [ProducesResponseType(typeof(RunExecutionResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status500InternalServerError)]
    public async Task<ActionResult<RunExecutionResponse>> Execute(
        Guid id,
        [FromBody] ExecuteFromDefinitionRequest req,
        CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(req.ValuationDate))
            return BadRequest("ValuationDate is required.");

        // Build a minimal NrtRunRequest — NrtService will merge definition fields
        var runRequest = new NrtRunRequest
        {
            DefinitionId  = id,
            ValuationDate = req.ValuationDate,
        };

        var result = await _nrtService.ExecuteRunAsync(runRequest, ct);
        return Ok(result);
    }
}

File: Xero.JavaWebApi/src/main/java/com/xero/webapi/controller/RunDefinitionsController.java
package com.xero.webapi.controller;

import com.xero.webapi.model.ExecuteFromDefinitionRequest;
import com.xero.webapi.model.NrtRunDefinitionDto;
import com.xero.webapi.model.NrtRunDefinitionSummary;
import com.xero.webapi.model.NrtRunRequest;
import com.xero.webapi.model.RunExecutionResponse;
import com.xero.webapi.model.SaveNrtRunDefinitionRequest;
import com.xero.webapi.service.NrtRunDefinitionService;
import com.xero.webapi.service.NrtService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.net.URI;
import java.time.LocalDate;
import java.time.format.DateTimeParseException;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * CRUD over saved {@code nrt_run_definitions} plus an "execute by id" shortcut.
 * Mirrors the .NET {@code RunDefinitionsController}.
 */
@RestController
@RequestMapping("/api/run-definitions")
public class RunDefinitionsController {

    private final NrtRunDefinitionService definitionService;
    private final NrtService              nrtService;

    public RunDefinitionsController(NrtRunDefinitionService definitionService,
                                    NrtService nrtService) {
        this.definitionService = definitionService;
        this.nrtService        = nrtService;
    }

    @GetMapping
    public List<NrtRunDefinitionSummary> getAll() {
        return definitionService.getAll();
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getById(@PathVariable UUID id) {
        NrtRunDefinitionDto dto = definitionService.getById(id);
        if (dto == null) {
            return ResponseEntity.status(404).body(Map.of("error", "Run definition " + id + " not found."));
        }
        return ResponseEntity.ok(dto);
    }

    @PostMapping
    public ResponseEntity<?> create(@Valid @RequestBody SaveNrtRunDefinitionRequest request) {
        if (isBlank(request.getName())) {
            return ResponseEntity.badRequest().body(Map.of("error", "Name is required."));
        }
        UUID id = definitionService.create(request);
        return ResponseEntity.created(URI.create("/api/run-definitions/" + id))
                .body(Map.of("definitionId", id));
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> update(@PathVariable UUID id,
                                    @Valid @RequestBody SaveNrtRunDefinitionRequest request) {
        if (isBlank(request.getName())) {
            return ResponseEntity.badRequest().body(Map.of("error", "Name is required."));
        }
        boolean updated = definitionService.update(id, request);
        if (!updated) {
            return ResponseEntity.status(404).body(Map.of("error", "Run definition " + id + " not found."));
        }
        return ResponseEntity.noContent().build();
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> delete(@PathVariable UUID id) {
        boolean removed = definitionService.delete(id);
        if (!removed) {
            return ResponseEntity.status(404).body(Map.of("error", "Run definition " + id + " not found."));
        }
        return ResponseEntity.noContent().build();
    }

    @PostMapping("/{id}/execute")
    public ResponseEntity<?> execute(@PathVariable UUID id,
                                     @RequestBody ExecuteFromDefinitionRequest request) {
        if (request == null || isBlank(request.getValuationDate())) {
            return ResponseEntity.badRequest().body(Map.of("error", "ValuationDate is required."));
        }

        LocalDate parsedDate;
        try {
            parsedDate = LocalDate.parse(request.getValuationDate());
        } catch (DateTimeParseException ex) {
            return ResponseEntity.badRequest()
                    .body(Map.of("error", "ValuationDate must be ISO-8601 (yyyy-MM-dd)."));
        }

        var runRequest = new NrtRunRequest();
        runRequest.setDefinitionId(id);
        runRequest.setValuationDate(parsedDate);

        RunExecutionResponse response = nrtService.executeRun(runRequest);
        return ResponseEntity.ok(response);
    }

    private static boolean isBlank(String s) {
        return s == null || s.isBlank();
    }
}
